### PR TITLE
JsSIP Socket interface

### PR DESCRIPTION
**Pull request by jmillan**
Make JsSIP socket agnostic.

The new JsSIP.Socket interface abstracts JsSIP from the mechanism used to send
and receive SIP traffic. JsSIP internal transport deals now with this interface
and hence, it is not attached to use the built-in WebSocket.

A new configuration parameter 'sockets' allows the user to provide it's own
sockets, which must implement the JsSIP.Socket interface.

The internal transport management and recovery handling has been
rewriten, leveraging the UA from any transport logic.

'hack_via_tcp' and 'hack_via_ws' configuration parameters have been removed.
The Via transport parameter is now seteable per Socket (see via_transport
Socket parameter).

### CODE REVIEW

File: lib/Socket.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    debugerror('Undefined JsSIP.Socket instance');
+    return false;
+  }
+
+  // Check Properties
+  try {
+    ['via_transport', 'url', 'sip_uri'].forEach(function(prop) {
+      if (!Utils.isString(socket[prop])) {
+        debugerror('Missing or invalid JsSIP.Socket property: '+ prop);
+        throw new Error();
```
Comment (saghul) [inline lib/Socket.js]: why throw and catch here instead of just returning false?

File: lib/Socket.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  } catch(e) {
+    return false;
+  }
+
+  // Check Methods
+  try {
+    ['connect', 'disconnect', 'send'].forEach(function(method) {
+      if (!Utils.isFunction(socket[method])) {
+        debugerror('Missing or invalid JsSIP.Socket method: '+ method);
+        throw new Error();
```
Comment (saghul) [inline lib/Socket.js]: ditto

File: lib/WebSocket.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+function WebSocket(url, node_ws_options) {
+  debug('new()');
+
+  var sip_uri = null;
+  var via_transport = null;
+
+  this.ws = null;
+
+  this.node_ws_options = node_ws_options || {};
```
Comment (saghul) [inline lib/WebSocket.js]: Object.assign({}, node_ws_options) ?

**Comment by jmillan** [comment]
@ibc, @saghul, consider giving it a quick look. The description speaks by itself. Documentation changes are already uploaded:

http://jssip.net/documentation/1.0.x/api/socket/
http://jssip.net/documentation/1.0.x/api/ua_configuration_parameters/#parameter_sockets

File: lib/Socket.js (hunk 1 above)
Comment (jmillan) [inline lib/Socket.js]: `return` inside a `foreEach` just continues on the next iteration

File: lib/Socket.js (hunk 2 above)
Comment (jmillan) [inline lib/Socket.js]: ditto :-)

**Comment by saghul** [comment]
I had a quick (not too deep) look at the code. This is great Jose!

Now that we are using semver, this would make it a 2.0.0, since we break backwards compatibility. But I like the change!

File: lib/WebSocket.js (hunk 3 above)
Comment (jmillan) [inline lib/WebSocket.js]: Nice :-). Just something to be applied gloally in a separate commit IMHO.

File: lib/Socket.js (hunk 1 above)
Comment (saghul) [inline lib/Socket.js]: oh, I had missed we were inside the `forEach`. Sorry for the noise!

**Comment by jmillan** [comment]
It doesn't break compatibility :-)

**Comment by jmillan** [comment]
BTW, fastest review ever 🍰

File: lib/WebSocket.js (hunk 3 above)
Comment (saghul) [inline lib/WebSocket.js]: Well, we can write new code The New Way (TM), right? Then adjust the rest. Gradual changes are always easier to make than gigantic ones.

**Comment by jmillan** [comment]
> It doesn't break compatibility :-)

Well, it does. Since hack_via_[tcp,udp] config parameters have been deprecated

**Comment by jmillan** [comment]
I perfectly agree on making it 2.0.0.

Lets wait for any comment from @ibc

**Comment by ibc** [comment]
@saghul `Object.assign()` is a ES6 feature not present in many browsers. We don't need to deal with that for now.

**Comment by ibc** [comment]
Good work @jmillan, some concerns coming to my mind:

* Does this break API so we would need to release 2.0.0?

* We could entirely remove the Node.js `websocket` module (which would avoid a native add-on building every time JsSIP is installed via `npm`) and, instead, provide a separate `jssip-node-websocket` module providing a `Socket` implementation just for Node.js.  

* May you provide a basic usage example (in browser with native WebSocket)? It's not clear for me whether the user needs to instantiate a WebSocket `Socket` instance or not in this scenario.

**Comment by ibc** [comment]
> `hack_via_[tcp,udp]` config parameters have been deprecated

Deprecated? or entirely removed?

**Comment by jmillan** [comment]
>>hack_via_[tcp,udp] config parameters have been deprecated

>Deprecated? or entirely removed?

Completely removed

**Comment by saghul** [comment]
> Completely removed

❤️

**Comment by saghul** [comment]
> Does this break API so we would need to release 2.0.0?
> 

Yep.

> We could entirely remove the Node.js websocket module (which would avoid a native add-on building every time JsSIP is installed via npm) and, instead, provide a separate jssip-node-websocket module providing a Socket implementation just for Node.js.
> 

Sounds good to me!

**Comment by jmillan** [comment]
> We could entirely remove the Node.js websocket module (which would avoid a native add-on building every time JsSIP is installed via npm) and, instead, provide a separate jssip-node-websocket module providing a Socket implementation just for Node.js.

Agreed. Let me check it a bit later 👍

**Comment by ibc** [comment]
@jmillan as I'm familiar with the node `websocket` module and its integration in JsSIP I may remove it stuff later once this PR is merged.

**Comment by jmillan** [comment]
Agreed @ibc,

The branch has been merged

**Comment by ibc** [comment]
yeah

**Comment by pjkundert** [comment]
I'd like to find / implement a plain TCP Transport for JsSIP.  I'm pretty sure I can't be the first one to want this, but I can't find any indication that anyone has done it!

If you know of any implementations of JsSIPP.Socket via plain TCP, could you let me know?  Thanks,
